### ISSUE TITLE
[apex] ApexCRUDViolation: Recognize User Mode in SOQL + DML

### ISSUE THREAD

**Issue by rsoesemann** [pmd/pmd#4146]
**Affects PMD Version:** 

**Rule:** [ApexCRUDViolation](https://pmd.github.io/latest/pmd_rules_apex_security.html#apexcrudviolation)

**Description:**
With the upcoming Winter '23 (API Version 56) Salesforce is going to add more native capabilities to enforce CRUD and FLS security in SOQL queries and DML statements as described here https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_enforce_usermode.htm.

Currently, the usage of such new features is ignored by PMD and marked as a CRUD Violation. Therefor the rule needs to be extended to recognize and handle them correctly.

**Code Sample demonstrating the issue:**

```java
public class UserMode {
	public void coverAllCasesWithTest() {
                // SOQL Queries cases
                
                Contact c;
                // Should be flagged a critical issue
                c = [SELECT Name FROM Contact];
                // Should be ignored
		c = [SELECT Name FROM Contact WITH USER_MODE];
                // Should be at best a warning because it ignores CRUD but explicitly
		c = [SELECT Name FROM Contact WITH SYSTEM_MODE];

                // DML cases

                // Should be flagged a critical issue
               insert contact;

                // Should be ignored
               insert as user contact;

                // Should be at best a warning because it ignores CRUD but explicitly
		insert as system contact;

               // ...and for ALL other occurrences of System.AccessLevel
	}
}
```

This issue should cover all cases with the optional accessLevel parameter. See [Dynamic SOQL](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_dynamic_soql.htm).
[Database.getQueryLocator methods](https://developer.salesforce.com/docs/atlas.en-us.240.0.apexref.meta/apexref/apex_methods_system_database.htm#apex_System_Database_getQueryLocator)
[Search.query methods](https://developer.salesforce.com/docs/atlas.en-us.240.0.apexref.meta/apexref/apex_methods_system_search.htm)
[Database DML methods](https://developer.salesforce.com/docs/atlas.en-us.240.0.apexref.meta/apexref/apex_methods_system_database.htm) (insert, update, upsert, merge, delete, undelete, and convertLead)
Includes the *Immediate and *Async methods, such as insertImmediate and deleteAsync.

This issuers should be easy to contribute as we can just look at how the related earlier enhancements were done: https://github.com/pmd/pmd/issues/2210

**Comment by Tarush-Singh35** [comment]
@rsoesemann I would like to contribute to this issue can you assign this issue to me?

**Comment by rsoesemann** [comment]
Thanks @adangel ;-)

**Comment by Tarush-Singh35** [comment]
yeah sure, I am getting started @rsoesemann we can work something out together. how can we connect @rbklaassen please let me know

**Comment by Tarush-Singh35** [comment]
@rbklaassen I am getting started with the issue if you want to join do ping me.

Thanks & Regards

**Comment by rbklaassen** [comment]
Hi @Tarush-Singh35 , go ahead and start on this. I've been quite busy last week and preparing a go-live at customer this week, so I don't think I can help out this or next week. I'll check in after that!

**Comment by Tarush-Singh35** [comment]
@rbklaassen through slack would be great

**Comment by rbklaassen** [comment]
I'm on slack with rklaassen@salesforce.com @Tarush-Singh35

**Comment by Tarush-Singh35** [comment]
> I'm on slack with [rklaassen@salesforce.com](mailto:rklaassen@salesforce.com) @Tarush-Singh35

yeah will connect with you through slack

**Comment by rsoesemann** [comment]
Any progress here,guys?

**Comment by rbklaassen** [comment]
Sorry, not yet. Have had COVID and a pretty busy schedule at customer.

**Comment by Tarush-Singh35** [comment]
I made some progress but I had an operation  so i was on rest I will raise the PR soon

**Comment by Tarush-Singh35** [comment]
@rbklaassen you ready to contribute?

**Comment by rbklaassen** [comment]
Yes, I can do some work on this tomorrow. Do you want to collaborate tomorrow @Tarush-Singh35 ?

**Comment by Tarush-Singh35** [comment]
Yeah @rbklaassen lets do this

**Comment by rbklaassen** [comment]
You need to connect with me through slack, I don't have enough details to connect to you @Tarush-Singh35

**Comment by Tarush-Singh35** [comment]
> You need to connect with me through slack, I don't have enough details to connect to you @Tarush-Singh35 

I will create a workspace and I will ask you to join it

**Comment by Tarush-Singh35** [comment]
Added you @rbklaassen in Slack

**Comment by rsoesemann** [comment]
@rbklaassen and @Tarush-Singh35 any progress here? Or do you have any blockers I can help with?

**Comment by rbklaassen** [comment]
Hi @rsoesemann, today we had a meeting together and discussed the way forward. I came to the conclusion that I miss the Java knowledge to really work on the code, but I can be the support-desk for @Tarush-Singh35 as I know Apex very good. Also I will be able to test his solution when needed on a local project for instance.

**Comment by Tarush-Singh35** [comment]
Yeah I am formulating a solution I think so I will be able to make a pull request soon @rsoesemann and @rbklaassen are mentoring me in the apex

**Comment by rsoesemann** [comment]
@Tarush-Singh35 I see two comments from @adangel which both make sense to me:

1. Explicit SYSTEM Mode should not detect an Issue. It's ok bahviour
2. Your code https://github.com/pmd/pmd/pull/4244/files#diff-32e0573a4d25115d3334e8f69cd4db0628e035eaa3831a90fc5f00a0e4e272f1R668 changes the logic also for other cases that relate to SECURITY ENFORCED. That's most probably the reason why existing test fail. Please run all tests before you submit a PR.

**Comment by adangel** [comment]
@rsoesemann Be aware, that the rule currently doesn't support query: see #2628 - so I didn't add this either.

Also, the currently available Jorje library (last updated in Feb. 2022) doesn't support the syntax `insert as user new Contact(LastName='foo')`. It fails with the following exception:

```
Caused by: apex.jorje.services.exception.ParseException: Syntax(error = UnexpectedToken(loc = (10, 16, 53, 55), token = 'as'))
```

There doesn't seem to be a new version available yet: https://github.com/forcedotcom/salesforcedx-vscode/commits/develop/packages/salesforcedx-vscode-apex/out/apex-jorje-lsp.jar

Apart from these two points, the PR #4244 fixes this issues.
